clang/test/Sema/builtin-counted-by-ref.c

```diff
@@ -12,8 +12,14 @@ struct non_fam_struct {
   short count;
 } *q;
 
```

Comment:
Clang does have support for `sized_by`, `counted_by_or_null`, and `sized_by_or_null` on "pointer" fields as well. What's the expected behavior for these cases? Should we emit an error for now that this is not yet supported? So later we can expand it to the right semantics.

```C
struct counted_buf {
  int count;
  int *buf __attribute__((counted_by(count)));
};

void t(struct counted_buf *p, int count) {
  if (__builtin_counted_by_ref(p->buf)) { // should it be an error that counted_by on pointer field not supported yet?
    *__builtin_counted_by_ref(p->buf) = count;
  }
}

struct sized_buf {
  int size;
  void *buf __attribute__((sized_by(size)));
};

void t(struct sized_buf *p, int size) {
  if (__builtin_counted_by_ref(p->buf)) { // should it be an error that sized_by is not supported yet?
    *__builtin_counted_by_ref(p->buf) = size;
  }
}

```

Reply:
We already require the argument to be a flexible array member. We can relax that when we get support for `counted_by` on pointers in structs. Though I suspect that the need for this builtin wouldn't really apply in such cases. :-)

Reply:
Nice! Could you please add the example in the test expected to be an error, if you haven't already?

Reply:
There are checks in clang/test/Sema/builtin-counted-by-ref.c. :-)

Reply:
@bwendling hmm though I don't see check for `counted_by` used on the pointer fields. To clarify, my examples are about `counted_by` exists on a "pointer" field. Which would be an error based on your implementation, but we need tests for it.